src/rules/noUnnecessaryBindRule.ts

```diff
@@ -0,0 +1,124 @@
+ * Copyright 2013 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tsutils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.TypedRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Prevents unnecessary and/or misleading scope bindings on functions.",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+            \`function\` expressions that are immediately bound to \`this\` are equivalent to \`() =>\` arrow lambdas.
+            Additionally, there's no use in binding a scope to an arrow lambda, as it already has one.
+        `,
+        requiresTypeInfo: true,
+        ruleName: "no-unnecessary-bind",
+        type: "style",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING_FUNCTION = "Don't bind \`this\` without arguments as a scope to a function. Use an arrow lambda instead.";
+
+    public static FAILURE_STRING_ARROW = "Don't bind scopes to arrow lambdas, as they already have a bound scope.";
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, undefined, program.getTypeChecker());
+    }
+}
+
+function walk(context: Lint.WalkContext<void>, typeChecker: ts.TypeChecker) {
+    function checkArrowFunction(node: ts.CallExpression, boundExpression: ts.Node): void {
+        if (node.arguments.length !== 1) {
+            return;
+        }
+
+        const replacement = Lint.Replacement.replaceNode(
+            node,
+            boundExpression.getText(context.sourceFile));
+
+        context.addFailureAtNode(node, Rule.FAILURE_STRING_ARROW, replacement);
+    }
+
+    function checkFunctionExpression(node: ts.CallExpression, valueDeclaration: ts.FunctionExpression): void {
+        if (node.arguments.length !== 1 || node.arguments[0].kind !== ts.SyntaxKind.ThisKeyword) {
+            return;
+        }
+
+        const parameters = valueDeclaration.parameters
+            .map((parameter) => parameter.getText(context.sourceFile))
+            .join(", ");
+        const body = valueDeclaration.body.getText(context.sourceFile);
+        const replacement = Lint.Replacement.replaceNode(
+            node,
+            `(${parameters}) => ${body}`);
```

Comment:
This fix will break code where the function expression is named and the name is used inside the function